File: bindings/ruby/lib/blazen/threed.rb
# frozen_string_literal: true

require "json"

module Blazen
  # Ruby wrapper for the +Compat3dProvider+ HTTP-proxy 3D pipeline
  # exposed by +blazen-3d+ via the +blazen-cabi+ surface (built behind the
  # +threed-compat-proxy+ feature flag — see
  # +crates/blazen-cabi/src/threed.rs+).
  #
  # The provider speaks +multipart/form-data+ to a configurable upstream
  # 3D server and exposes four async verbs:
  #
  # * {Compat3dProvider#texturize} — apply / generate a PBR texture for
  #   an input mesh GLB
  # * {Compat3dProvider#rig} — auto-rig a mesh (armature + optional
  #   skin weights)
  # * {Compat3dProvider#refine} — decimate, hole-fill, UV-unwrap,
  #   retopologize, smooth a mesh
  # * {Compat3dProvider#animate} — produce an animated GLB from a text
  #   prompt, MP4 driving video, or BVH motion clip
  #
  # All four methods compose with +Fiber.scheduler+ when one is active
  # (via {Blazen::FFI.await_future}) so they cooperate with the +async+
  # gem; the matching +*_blocking+ siblings drive the future on the
  # cabi tokio runtime when no scheduler is in scope.
  #
  # @example basic usage
  #   provider = Blazen::ThreeD::Compat3dProvider.new(
  #     base_url: "https://3d.example.com",
  #     api_key: ENV["THREED_API_KEY"],
  #   )
  #   mesh = File.binread("input.glb")
  #   result = provider.texturize(
  #     mesh,
  #     prompt: "weathered bronze",
  #     pbr: true,
  #   )
  #   File.binwrite("textured.glb", result.glb_bytes)
  module ThreeD
    # Raised when the native library was built without the
    # +threed-compat-proxy+ feature so the Compat3d symbols are missing.
    class FeatureMissing < Blazen::UnsupportedError
      def initialize(msg = nil)
        super(msg || "Blazen::ThreeD requires libblazen_cabi built with the " \
                     "`threed-compat-proxy` feature; current build omits it.")
      end
    end

    # @return [Boolean] whether the native lib supports the Compat3d
    #   surface in this build.
    def self.available?
      Blazen::FFI.respond_to?(:threed_available?) && Blazen::FFI.threed_available?
    end

    # Result of any of the four pipeline stages. Wraps the opaque
    # +BlazenCompat3dResult *+ pointer; all stage-specific accessors
    # gracefully return +nil+ / empty collections when called on the
    # wrong stage's result.
    class Result
      # @return [::FFI::AutoPointer] underlying +BlazenCompat3dResult *+
      attr_reader :ptr
      # @return [Symbol] one of +:texturize+, +:rig+, +:refine+,
      #   +:animate+ — set by the calling verb so stage-specific
      #   accessors can short-circuit on the wrong variant.
      attr_reader :stage

      # @api private
      def initialize(raw_ptr, stage)
        if raw_ptr.nil? || raw_ptr.null?
          raise Blazen::InternalError,
                "Blazen::ThreeD::Result: native handle is null"
        end

        @stage = stage
        @ptr = ::FFI::AutoPointer.new(
          raw_ptr, Blazen::FFI.method(:blazen_compat3d_result_free),
        )
      end

      # @return [String] the produced GLB bytes (binary-encoded)
      def glb_bytes
        out_ptr = ::FFI::MemoryPointer.new(:pointer)
        out_len = ::FFI::MemoryPointer.new(:size_t)
        rc = Blazen::FFI.blazen_compat3d_result_glb_bytes(@ptr, out_ptr, out_len)
        raise Blazen::InternalError, "glb_bytes accessor returned #{rc}" unless rc.zero?

        data_ptr = out_ptr.read_pointer
        length   = out_len.read(:size_t)
        return String.new(encoding: Encoding::BINARY) if data_ptr.nil? || data_ptr.null? || length.zero?

        data_ptr.read_bytes(length).force_encoding(Encoding::BINARY)
      end

      # @return [String] the MIME type of {#glb_bytes}; always
      #   +"model/gltf-binary"+ from this backend.
      def mime_type
        Blazen::FFI.consume_cstring(Blazen::FFI.blazen_compat3d_result_mime_type(@ptr)) || ""
      end

      # @return [Boolean] +true+ when this is a +texturize+ result and
      #   the upstream returned an out-of-band PBR-map bundle.
      def pbr_maps?
        Blazen::FFI.blazen_compat3d_result_has_pbr_maps(@ptr) == 1
      end

      # Returns the bytes of one PBR channel. Returns +nil+ for any
      # other stage / missing channel.
      #
      # @param channel [Symbol] one of +:albedo+, +:normal+,
      #   +:roughness+, +:metallic+
      # @return [String, nil]
      def pbr_map(channel)
        ch = pbr_channel_id(channel)
        out_ptr = ::FFI::MemoryPointer.new(:pointer)
        out_len = ::FFI::MemoryPointer.new(:size_t)
        rc = Blazen::FFI.blazen_compat3d_result_pbr_map_bytes(@ptr, ch, out_ptr, out_len)
        return nil unless rc.zero?

        length   = out_len.read(:size_t)
        data_ptr = out_ptr.read_pointer
        return nil if data_ptr.nil? || data_ptr.null? || length.zero?

        data_ptr.read_bytes(length).force_encoding(Encoding::BINARY)
      end

      # @return [Array<String>] bone names for a +rig+ result (in
      #   depth-first traversal order); empty array for any other stage.
      def bone_names
        len = Blazen::FFI.blazen_compat3d_result_bone_names_count(@ptr)
        return [] if len.zero?

        Array.new(len) do |i|
          Blazen::FFI.consume_cstring(
            Blazen::FFI.blazen_compat3d_result_bone_name_get(@ptr, i),
          ) || ""
        end
      end

      # @return [Hash{Symbol => Integer, nil}] +{input_tri_count:,
      #   output_tri_count:, uv_chart_count:}+ for a +refine+ result;
      #   +uv_chart_count+ is +nil+ when UV unwrapping wasn't requested.
      #   Returns +nil+ for any other stage.
      def refine_stats
        return nil unless stage == :refine

        uv = Blazen::FFI.blazen_compat3d_result_refine_uv_chart_count(@ptr)
        {
          input_tri_count: Blazen::FFI.blazen_compat3d_result_refine_input_tri_count(@ptr),
          output_tri_count: Blazen::FFI.blazen_compat3d_result_refine_output_tri_count(@ptr),
          uv_chart_count: uv.negative? ? nil : uv,
        }
      end

      # @return [Float, nil] the produced animation duration in seconds
      #   for an +animate+ result; +nil+ for any other stage.
      def duration_seconds
        return nil unless stage == :animate

        Blazen::FFI.blazen_compat3d_result_animate_duration_seconds(@ptr)
      end

      # @return [Integer, nil] the produced animation framerate for an
      #   +animate+ result; +nil+ for any other stage.
      def fps
        return nil unless stage == :animate

        Blazen::FFI.blazen_compat3d_result_animate_fps(@ptr)
      end

      private

      def pbr_channel_id(channel)
        case channel
        when :albedo    then Blazen::FFI::PBR_MAP_ALBEDO
        when :normal    then Blazen::FFI::PBR_MAP_NORMAL
        when :roughness then Blazen::FFI::PBR_MAP_ROUGHNESS
        when :metallic  then Blazen::FFI::PBR_MAP_METALLIC
        else
          raise Blazen::ValidationError,
                "unknown PBR channel: #{channel.inspect} " \
                "(expected :albedo, :normal, :roughness, :metallic)"
        end
      end
    end

    # HTTP-proxy 3D provider. See {Blazen::ThreeD} for usage.
    class Compat3dProvider
      # @return [::FFI::AutoPointer] underlying +BlazenCompat3dProvider *+
      attr_reader :ptr

      # @param base_url [String] upstream root URL
      #   (e.g. +"https://3d.example.com"+)
      # @param api_key [String, nil] optional bearer token attached as
      #   +Authorization: Bearer ...+
      # @param timeout_secs [Integer, nil] optional per-request timeout
      #   in seconds. +nil+ / +0+ → 10-minute default.
      def initialize(base_url:, api_key: nil, timeout_secs: nil)
        unless Blazen::ThreeD.available?
          raise FeatureMissing
        end
        raise Blazen::ValidationError, "base_url must be a non-empty string" \
          if base_url.nil? || base_url.to_s.empty?

        raw = Blazen::FFI.with_cstring(base_url.to_s) do |base|
          Blazen::FFI.with_cstring(api_key) do |key|
            Blazen::FFI.blazen_compat3d_provider_new(
              base, key, Integer(timeout_secs || 0),
            )
          end
        end
        if raw.nil? || raw.null?
          raise Blazen::InternalError,
                "blazen_compat3d_provider_new returned null"
        end

        @ptr = ::FFI::AutoPointer.new(
          raw, Blazen::FFI.method(:blazen_compat3d_provider_free),
        )
      end

      # Apply or generate a texture / material for an input mesh GLB.
      #
      # @param mesh_glb [String] input mesh as GLB bytes (binary)
      # @param prompt [String, nil] text-guided texture prompt
      # @param reference_image [String, nil] PNG/JPEG reference bytes
      # @param style [String, nil] backend-specific style preset
      # @param resolution [Integer, nil] target square texture
      #   resolution in pixels
      # @param pbr [Boolean] request a full PBR material bundle
      # @return [Blazen::ThreeD::Result]
      # @raise [Blazen::Error] on any provider / network failure
      def texturize(mesh_glb, prompt: nil, reference_image: nil, style: nil,
                    resolution: nil, pbr: false)
        json = compact_json(
          prompt: prompt, style: style, resolution: resolution, pbr: pbr,
        )
        with_byte_buffer(mesh_glb) do |mesh_ptr, mesh_len|
          with_byte_buffer(reference_image) do |ref_ptr, ref_len|
            Blazen::FFI.with_cstring(json) do |json_ptr|
              fut = Blazen::FFI.blazen_compat3d_texturize(
                @ptr, mesh_ptr, mesh_len, ref_ptr, ref_len, json_ptr,
              )
              take_future_result(fut, :texturize)
            end
          end
        end
      end

      # Auto-rig a 3D mesh.
      #
      # @param mesh_glb [String]
      # @param template [String, nil] +"humanoid"+ / +"quadruped"+ /
      #   +"auto"+ (backend-specific)
      # @param skin [Boolean] apply skin-weight painting after armature
      # @param pose_hint [String, nil] +"t-pose"+ / +"a-pose"+ / backend
      #   JSON
      # @return [Blazen::ThreeD::Result]
      def rig(mesh_glb, template: nil, skin: false, pose_hint: nil)
        json = compact_json(template: template, skin: skin, pose_hint: pose_hint)
        with_byte_buffer(mesh_glb) do |mesh_ptr, mesh_len|
          Blazen::FFI.with_cstring(json) do |json_ptr|
            fut = Blazen::FFI.blazen_compat3d_rig(
              @ptr, mesh_ptr, mesh_len, json_ptr,
            )
            take_future_result(fut, :rig)
          end
        end
      end

      # Refine a mesh: decimate / hole-fill / UV-unwrap / retopologize /
      # smooth.
      #
      # @param mesh_glb [String]
      # @param decimate_target_tris [Integer, nil]
      # @param fill_holes [Boolean]
      # @param unwrap_uvs [Boolean]
      # @param retopologize [Boolean]
      # @param smooth_iterations [Integer, nil]
      # @return [Blazen::ThreeD::Result]
      def refine(mesh_glb, decimate_target_tris: nil, fill_holes: false,
                 unwrap_uvs: false, retopologize: false, smooth_iterations: nil)
        json = compact_json(
          decimate_target_tris: decimate_target_tris,
          fill_holes: fill_holes,
          unwrap_uvs: unwrap_uvs,
          retopologize: retopologize,
          smooth_iterations: smooth_iterations,
        )
        with_byte_buffer(mesh_glb) do |mesh_ptr, mesh_len|
          Blazen::FFI.with_cstring(json) do |json_ptr|
            fut = Blazen::FFI.blazen_compat3d_refine(
              @ptr, mesh_ptr, mesh_len, json_ptr,
            )
            take_future_result(fut, :refine)
          end
        end
      end

      # Animate a rigged mesh from a text prompt, MP4 video, or BVH clip.
      #
      # @param rigged_glb [String] rigged input mesh bytes
      # @param prompt [String, nil] text motion prompt
      # @param driving_video [String, nil] MP4 driving video bytes
      # @param bvh_motion [String, nil] BVH motion-capture clip bytes
      # @param duration_seconds [Float, nil]
      # @param fps [Integer, nil]
      # @param loop_animation [Boolean] mark output as a seamless loop
      # @return [Blazen::ThreeD::Result]
      def animate(rigged_glb, prompt: nil, driving_video: nil, bvh_motion: nil,
                  duration_seconds: nil, fps: nil, loop_animation: false)
        json = compact_json(
          prompt: prompt, duration_seconds: duration_seconds, fps: fps,
          loop_animation: loop_animation,
        )
        with_byte_buffer(rigged_glb) do |mesh_ptr, mesh_len|
          with_byte_buffer(driving_video) do |drv_ptr, drv_len|
            with_byte_buffer(bvh_motion) do |bvh_ptr, bvh_len|
              Blazen::FFI.with_cstring(json) do |json_ptr|
                fut = Blazen::FFI.blazen_compat3d_animate(
                  @ptr, mesh_ptr, mesh_len, drv_ptr, drv_len,
                  bvh_ptr, bvh_len, json_ptr,
                )
                take_future_result(fut, :animate)
              end
            end
          end
        end
      end

      private

      # Builds a JSON string with +nil+ values dropped and +false+
      # booleans normalised. Mirrors the cbindgen wire-format
      # contract — see +crates/blazen-cabi/src/threed.rs+ for the
      # accepted shape per stage.
      def compact_json(fields)
        compact = fields.reject { |_, v| v.nil? }
        JSON.generate(compact)
      end

      # Yields a +(MemoryPointer, length)+ pair for +bytes+. Yields
      # +(nil, 0)+ when +bytes+ is nil or empty. The buffer's strong
      # ref is held in the block scope so GC can't reclaim it
      # mid-call.
      def with_byte_buffer(bytes)
        if bytes.nil? || bytes.empty?
          return yield(nil, 0)
        end

        raw = bytes.is_a?(String) ? bytes.b : bytes.pack("C*")
        len = raw.bytesize
        buf = ::FFI::MemoryPointer.new(:uint8, len)
        buf.write_bytes(raw, 0, len)
        yield(buf, len)
      end

      def take_future_result(fut, stage)
        if fut.nil? || fut.null?
          raise Blazen::ValidationError,
                "blazen_compat3d_#{stage} rejected the call (null inputs?)"
        end

        Blazen::FFI.await_future(fut) do |f|
          out = ::FFI::MemoryPointer.new(:pointer)
          out_err = ::FFI::MemoryPointer.new(:pointer)
          Blazen::FFI.blazen_future_take_compat3d_result(f, out, out_err)
          Blazen::FFI.check_error!(out_err)
          Result.new(out.read_pointer, stage)
        end
      end
    end
  end
end

File: crates/blazen-uniffi/src/threed.rs
//! UniFFI bindings for the [`blazen_3d`] crate's HTTP-proxy backend.
//!
//! Exposes the four 3D-pipeline capability traits — texturizer, rigger,
//! refiner, animator — as a single [`Compat3dProvider`] UniFFI object
//! whose four async methods forward to a configurable upstream service
//! over `multipart/form-data`. Mirrors the surface in
//! `crates/blazen-py/src/threed.rs` so the Go / Swift / Kotlin / Ruby
//! bindings get the same 3D-pipeline shape as the Python binding.
//!
//! ## Surface
//!
//! * Request records: [`TexturizeRequest`], [`RigRequest`],
//!   [`RefineRequest`], [`AnimateRequest`].
//! * Result records: [`TexturizeResult`] (with [`PbrMaps`]), [`RigResult`],
//!   [`RefineResult`] (with [`RefineStats`]), [`AnimateResult`].
//! * Provider object: [`Compat3dProvider`] with async methods
//!   `texturize`, `rig`, `refine`, `animate`.
//! * Error: [`ThreeDError`] — a flat error enum that absorbs the four
//!   parallel error enums from `blazen-3d` (`Texturizer3dError`,
//!   `Rigger3dError`, `Refiner3dError`, `Animator3dError`).
//!
//! UniFFI prefers types defined in the crate that exports them, so the
//! request and result types are local mirrors of the foreign
//! `blazen_3d::*` types with `From` conversions at the seam.

use std::sync::Arc;
use std::time::Duration;

use blazen_3d::backends::compat::Compat3dProvider as CoreCompat3dProvider;
use blazen_3d::{
    AnimateRequest as CoreAnimateRequest, AnimateResult as CoreAnimateResult, Animator3dBackend,
    Animator3dError, PbrMaps as CorePbrMaps, RefineRequest as CoreRefineRequest,
    RefineResult as CoreRefineResult, RefineStats as CoreRefineStats, Refiner3dBackend,
    Refiner3dError, RigRequest as CoreRigRequest, RigResult as CoreRigResult, Rigger3dBackend,
    Rigger3dError, TexturizeRequest as CoreTexturizeRequest,
    TexturizeResult as CoreTexturizeResult, Texturizer3dBackend, Texturizer3dError,
};
use thiserror::Error;

// ---------------------------------------------------------------------------
// Error type
// ---------------------------------------------------------------------------

/// Canonical error returned by every [`Compat3dProvider`] method.
///
/// Absorbs the four parallel error enums from [`blazen_3d`] —
/// [`Texturizer3dError`], [`Rigger3dError`], [`Refiner3dError`], and
/// [`Animator3dError`] — into one flat shape. Foreign callers switch
/// on the variant rather than on which stage produced the error
/// (the call site already determines that).
#[derive(Debug, Error, uniffi::Error)]
pub enum ThreeDError {
    /// The active backend reported a runtime failure (HTTP error,
    /// inference error, etc.).
    #[error("backend: {message}")]
    Backend { message: String },

    /// The caller-supplied input was malformed — invalid mesh bytes,
    /// unsupported container format, malformed request fields, etc.
    #[error("invalid input: {message}")]
    InvalidInput { message: String },

    /// I/O failure while reading a mesh file, reference image, or
    /// model file.
    #[error("io: {message}")]
    Io { message: String },

    /// The selected backend is not available in this build (e.g. the
    /// `compat-proxy` feature is disabled).
    #[error("engine not available: {message}")]
    EngineNotAvailable { message: String },

    /// The capability requested is not supported by the active
    /// backend (e.g. PBR maps from an albedo-only texturizer, video-
    /// driven motion on a text-only animator).
    #[error("unsupported: {message}")]
    Unsupported { message: String },
}

impl From<Texturizer3dError> for ThreeDError {
    fn from(err: Texturizer3dError) -> Self {
        match err {
            Texturizer3dError::Io(e) => Self::Io {
                message: e.to_string(),
            },
            Texturizer3dError::EngineNotAvailable(m) => Self::EngineNotAvailable { message: m },
            Texturizer3dError::InvalidInput(m) => Self::InvalidInput { message: m },
            Texturizer3dError::Backend(m) => Self::Backend { message: m },
            Texturizer3dError::Unsupported(m) => Self::Unsupported { message: m },
        }
    }
}

impl From<Rigger3dError> for ThreeDError {
    fn from(err: Rigger3dError) -> Self {
        match err {
            Rigger3dError::Io(e) => Self::Io {
                message: e.to_string(),
            },
            Rigger3dError::EngineNotAvailable(m) => Self::EngineNotAvailable { message: m },
            Rigger3dError::InvalidInput(m) => Self::InvalidInput { message: m },
            Rigger3dError::Backend(m) => Self::Backend { message: m },
            Rigger3dError::Unsupported(m) => Self::Unsupported { message: m },
        }
    }
}

impl From<Refiner3dError> for ThreeDError {
    fn from(err: Refiner3dError) -> Self {
        match err {
            Refiner3dError::Io(e) => Self::Io {
                message: e.to_string(),
            },
            Refiner3dError::EngineNotAvailable(m) => Self::EngineNotAvailable { message: m },
            Refiner3dError::InvalidInput(m) => Self::InvalidInput { message: m },
            Refiner3dError::Backend(m) => Self::Backend { message: m },
            Refiner3dError::Unsupported(m) => Self::Unsupported { message: m },
        }
    }
}

impl From<Animator3dError> for ThreeDError {
    fn from(err: Animator3dError) -> Self {
        match err {
            Animator3dError::Io(e) => Self::Io {
                message: e.to_string(),
            },
            Animator3dError::EngineNotAvailable(m) => Self::EngineNotAvailable { message: m },
            Animator3dError::InvalidInput(m) => Self::InvalidInput { message: m },
            Animator3dError::Backend(m) => Self::Backend { message: m },
            Animator3dError::Unsupported(m) => Self::Unsupported { message: m },
        }
    }
}

// ---------------------------------------------------------------------------
// PbrMaps / RefineStats
// ---------------------------------------------------------------------------

/// Bundle of PBR (physically-based rendering) material maps produced by
/// a texturizer backend. `albedo_png` is always populated; the other
/// channels are optional and depend on what the backend produces.
#[derive(Debug, Clone, uniffi::Record)]
pub struct PbrMaps {
    /// Base-color / diffuse texture as PNG bytes. Always present.
    pub albedo_png: Vec<u8>,
    /// Tangent-space normal map as PNG bytes, if produced.
    pub normal_png: Option<Vec<u8>>,
    /// Linear roughness map as PNG bytes, if produced.
    pub roughness_png: Option<Vec<u8>>,
    /// Linear metallic map as PNG bytes, if produced.
    pub metallic_png: Option<Vec<u8>>,
}

impl From<CorePbrMaps> for PbrMaps {
    fn from(m: CorePbrMaps) -> Self {
        Self {
            albedo_png: m.albedo_png,
            normal_png: m.normal_png,
            roughness_png: m.roughness_png,
            metallic_png: m.metallic_png,
        }
    }
}

impl From<PbrMaps> for CorePbrMaps {
    fn from(m: PbrMaps) -> Self {
        Self {
            albedo_png: m.albedo_png,
            normal_png: m.normal_png,
            roughness_png: m.roughness_png,
            metallic_png: m.metallic_png,
        }
    }
}

/// Summary statistics emitted by a [`Compat3dProvider::refine`] call.
#[derive(Debug, Clone, uniffi::Record)]
pub struct RefineStats {
    /// Triangle count of the input mesh.
    pub input_tri_count: u32,
    /// Triangle count of the output (refined) mesh.
    pub output_tri_count: u32,
    /// When UV unwrapping ran, the number of UV charts the unwrap
    /// produced. `None` when UV unwrapping did not run for this call.
    pub uv_chart_count: Option<u32>,
}

impl From<CoreRefineStats> for RefineStats {
    fn from(s: CoreRefineStats) -> Self {
        Self {
            input_tri_count: s.input_tri_count,
            output_tri_count: s.output_tri_count,
            uv_chart_count: s.uv_chart_count,
        }
    }
}

impl From<RefineStats> for CoreRefineStats {
    fn from(s: RefineStats) -> Self {
        Self {
            input_tri_count: s.input_tri_count,
            output_tri_count: s.output_tri_count,
            uv_chart_count: s.uv_chart_count,
        }
    }
}

// ---------------------------------------------------------------------------
// TexturizeRequest / TexturizeResult
// ---------------------------------------------------------------------------

/// Request parameters for [`Compat3dProvider::texturize`].
#[derive(Debug, Clone, uniffi::Record)]
pub struct TexturizeRequest {
    /// Text-guided texture prompt (e.g. `"weathered bronze"`).
    pub prompt: Option<String>,
    /// Image-guided reference (PNG or JPEG bytes) used as a style anchor.
    pub reference_image: Option<Vec<u8>>,
    /// Backend-specific style preset (`"stylized"`, `"realistic"`, ...).
    pub style: Option<String>,
    /// Target square texture resolution in pixels.
    pub resolution: Option<u32>,
    /// `true` to request a full PBR material bundle.
    pub pbr: bool,
}

impl From<TexturizeRequest> for CoreTexturizeRequest {
    fn from(r: TexturizeRequest) -> Self {
        Self {
            prompt: r.prompt,
            reference_image: r.reference_image,
            style: r.style,
            resolution: r.resolution,
            pbr: r.pbr,
        }
    }
}

/// Result of a successful [`Compat3dProvider::texturize`] call.
#[derive(Debug, Clone, uniffi::Record)]
pub struct TexturizeResult {
    /// GLB bytes with the new texture (and PBR maps if any) embedded.
    pub textured_glb: Vec<u8>,
    /// MIME type of `textured_glb`; always `"model/gltf-binary"`.
    pub mime_type: String,
    /// Optional out-of-band PBR map bundle. Duplicates of the maps
    /// embedded in `textured_glb` when present.
    pub pbr_maps: Option<PbrMaps>,
}

impl From<CoreTexturizeResult> for TexturizeResult {
    fn from(r: CoreTexturizeResult) -> Self {
        Self {
            textured_glb: r.textured_glb,
            mime_type: r.mime_type,
            pbr_maps: r.pbr_maps.map(PbrMaps::from),
        }
    }
}

// ---------------------------------------------------------------------------
// RigRequest / RigResult
// ---------------------------------------------------------------------------

/// Request parameters for [`Compat3dProvider::rig`].
#[derive(Debug, Clone, uniffi::Record)]
pub struct RigRequest {
    /// Target rig template (`"humanoid"`, `"quadruped"`, `"auto"`).
    pub template: Option<String>,
    /// `true` to apply skin-weight painting after armature placement.
    pub skin: bool,
    /// Optional pose hint (`"t-pose"`, `"a-pose"`, or backend-specific JSON).
    pub pose_hint: Option<String>,
}

impl From<RigRequest> for CoreRigRequest {
    fn from(r: RigRequest) -> Self {
        Self {
            template: r.template,
            skin: r.skin,
            pose_hint: r.pose_hint,
        }
    }
}

/// Result of a successful [`Compat3dProvider::rig`] call.
#[derive(Debug, Clone, uniffi::Record)]
pub struct RigResult {
    /// GLB bytes with the new armature (and skin weights, if requested) embedded.
    pub rigged_glb: Vec<u8>,
    /// MIME type of `rigged_glb`; always `"model/gltf-binary"`.
    pub mime_type: String,
    /// Names of bones in the produced armature, in depth-first traversal order.
    pub bone_names: Vec<String>,
}

impl From<CoreRigResult> for RigResult {
    fn from(r: CoreRigResult) -> Self {
        Self {
            rigged_glb: r.rigged_glb,
            mime_type: r.mime_type,
            bone_names: r.bone_names,
        }
    }
}

// ---------------------------------------------------------------------------
// RefineRequest / RefineResult
// ---------------------------------------------------------------------------

/// Request parameters for [`Compat3dProvider::refine`].
#[derive(Debug, Clone, uniffi::Record)]
pub struct RefineRequest {
    /// Decimate the mesh towards this triangle count.
    pub decimate_target_tris: Option<u32>,
    /// `true` to fill holes via screened poisson reconstruction.
    pub fill_holes: bool,
    /// `true` to compute a new UV unwrap.
    pub unwrap_uvs: bool,
    /// `true` to retopologize the mesh.
    pub retopologize: bool,
    /// Laplacian / Taubin smoothing iteration count.
    pub smooth_iterations: Option<u32>,
}

impl From<RefineRequest> for CoreRefineRequest {
    fn from(r: RefineRequest) -> Self {
        Self {
            decimate_target_tris: r.decimate_target_tris,
            fill_holes: r.fill_holes,
            unwrap_uvs: r.unwrap_uvs,
            retopologize: r.retopologize,
            smooth_iterations: r.smooth_iterations,
        }
    }
}

/// Result of a successful [`Compat3dProvider::refine`] call.
#[derive(Debug, Clone, uniffi::Record)]
pub struct RefineResult {
    /// GLB bytes with the requested refinement passes applied.
    pub refined_glb: Vec<u8>,
    /// MIME type of `refined_glb`; always `"model/gltf-binary"`.
    pub mime_type: String,
    /// Before/after statistics for the refinement run.
    pub stats: RefineStats,
}

impl From<CoreRefineResult> for RefineResult {
    fn from(r: CoreRefineResult) -> Self {
        Self {
            refined_glb: r.refined_glb,
            mime_type: r.mime_type,
            stats: RefineStats::from(r.stats),
        }
    }
}

// ---------------------------------------------------------------------------
// AnimateRequest / AnimateResult
// ---------------------------------------------------------------------------

/// Request parameters for [`Compat3dProvider::animate`].
#[derive(Debug, Clone, uniffi::Record)]
pub struct AnimateRequest {
    /// Text-guided motion prompt (e.g. `"walks forward and waves"`).
    pub prompt: Option<String>,
    /// Optional MP4 bytes for video-driven motion transfer.
    pub driving_video: Option<Vec<u8>>,
    /// Optional BVH motion-capture clip bytes.
    pub bvh_motion: Option<Vec<u8>>,
    /// Requested animation duration in seconds.
    pub duration_seconds: Option<f32>,
    /// Requested animation framerate.
    pub fps: Option<u32>,
    /// `true` to mark the produced animation as a seamless loop.
    pub loop_animation: bool,
}

impl From<AnimateRequest> for CoreAnimateRequest {
    fn from(r: AnimateRequest) -> Self {
        Self {
            prompt: r.prompt,
            driving_video: r.driving_video,
            bvh_motion: r.bvh_motion,
            duration_seconds: r.duration_seconds,
            fps: r.fps,
            loop_animation: r.loop_animation,
        }
    }
}

/// Result of a successful [`Compat3dProvider::animate`] call.
#[derive(Debug, Clone, uniffi::Record)]
pub struct AnimateResult {
    /// GLB bytes with the animation track(s) embedded.
    pub animated_glb: Vec<u8>,
    /// MIME type of `animated_glb`; always `"model/gltf-binary"`.
    pub mime_type: String,
    /// Actual produced duration in seconds (may differ from the request).
    pub duration_seconds: f32,
    /// Actual produced framerate in frames per second (may differ from the request).
    pub fps: u32,
}

impl From<CoreAnimateResult> for AnimateResult {
    fn from(r: CoreAnimateResult) -> Self {
        Self {
            animated_glb: r.animated_glb,
            mime_type: r.mime_type,
            duration_seconds: r.duration_seconds,
            fps: r.fps,
        }
    }
}

// ---------------------------------------------------------------------------
// Compat3dProvider
// ---------------------------------------------------------------------------

/// HTTP-proxy backend implementing all four 3D-pipeline capability
/// traits against a configurable upstream service.
///
/// For every stage, this provider POSTs a `multipart/form-data` request
/// with the mesh GLB and a JSON request body to
/// `{base_url}/v1/3d/{texturize,rig,refine,animate}`, and decodes a
/// base64-wrapped JSON response into the corresponding result record.
#[derive(uniffi::Object)]
pub struct Compat3dProvider {
    inner: Arc<CoreCompat3dProvider>,
}

#[uniffi::export(async_runtime = "tokio")]
impl Compat3dProvider {
    /// Construct a new HTTP-proxy provider.
    ///
    /// `base_url` is the upstream root URL (e.g.
    /// `"https://3d.example.com"`). `api_key` is an optional bearer
    /// token attached as `Authorization: Bearer ...`. `timeout_secs`
    /// is an optional per-request timeout in seconds (default 600).
    #[uniffi::constructor]
    #[must_use]
    pub fn new(base_url: String, api_key: Option<String>, timeout_secs: Option<u32>) -> Arc<Self> {
        let mut provider = CoreCompat3dProvider::new(base_url);
        if let Some(key) = api_key {
            provider = provider.with_api_key(key);
        }
        if let Some(secs) = timeout_secs {
            provider = provider.with_timeout(Duration::from_secs(u64::from(secs)));
        }
        Arc::new(Self {
            inner: Arc::new(provider),
        })
    }

    /// Apply or generate a texture/material for an existing 3D mesh.
    pub async fn texturize(
        self: Arc<Self>,
        mesh_glb: Vec<u8>,
        request: TexturizeRequest,
    ) -> Result<TexturizeResult, ThreeDError> {
        let provider = Arc::clone(&self.inner);
        let core_req = CoreTexturizeRequest::from(request);
        let result = provider.texturize(&mesh_glb, core_req).await?;
        Ok(TexturizeResult::from(result))
    }

    /// Auto-rig a 3D mesh, producing a GLB with skeletal armature and
    /// (optionally) skin weights embedded.
    pub async fn rig(
        self: Arc<Self>,
        mesh_glb: Vec<u8>,
        request: RigRequest,
    ) -> Result<RigResult, ThreeDError> {
        let provider = Arc::clone(&self.inner);
        let core_req = CoreRigRequest::from(request);
        let result = provider.rig(&mesh_glb, core_req).await?;
        Ok(RigResult::from(result))
    }

    /// Refine a 3D mesh: decimate, fill holes, unwrap UVs, retopologize, smooth.
    pub async fn refine(
        self: Arc<Self>,
        mesh_glb: Vec<u8>,
        request: RefineRequest,
    ) -> Result<RefineResult, ThreeDError> {
        let provider = Arc::clone(&self.inner);
        let core_req = CoreRefineRequest::from(request);
        let result = provider.refine(&mesh_glb, core_req).await?;
        Ok(RefineResult::from(result))
    }

    /// Animate a rigged 3D mesh from a text prompt, motion-capture clip,
    /// or driving video.
    pub async fn animate(
        self: Arc<Self>,
        rigged_glb: Vec<u8>,
        request: AnimateRequest,
    ) -> Result<AnimateResult, ThreeDError> {
        let provider = Arc::clone(&self.inner);
        let core_req = CoreAnimateRequest::from(request);
        let result = provider.animate(&rigged_glb, core_req).await?;
        Ok(AnimateResult::from(result))
    }
}
